src/expr/src/expr/expr_array_positions.rs

```diff
@@ -0,0 +1,179 @@
+    /// given as first argument. Note the behavior is slightly different from PG.
+    ///
+    /// Examples:
+    ///
+    /// ```slt
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[1]);
+    /// ----
+    /// {1}
+    ///
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[2]);
+    /// ----
+    /// {2,4}
+    ///
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], null::int[]);
+    /// ----
+    /// {5}
+    ///
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[4]);
+    /// ----
+    /// {}
+    ///
+    /// query T
+    /// select array_positions(null::int[], 1);
+    /// ----
+    /// NULL
+    ///
+    /// query T
+    /// select array_positions(ARRAY[array[1],array[2],array[3],array[2],null::int[]], array[3.14]);
+    /// ----
+    /// {}
+    ///
+    /// query T
+    /// select array_positions(array[1,NULL,NULL,3], NULL::int);
+    /// ----
+    /// {2,3}
+    ///
+    /// statement error
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], 1);
+    ///
+    /// statement error
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[array[3]]);
+    ///
+    /// statement error
+    /// select array_positions(ARRAY[array[1],array[2],array[3],array[2],null::int[]], array[true]);
+    /// ```
+    fn evaluate(left: DatumRef<'_>, right: DatumRef<'_>) -> Datum {
+        match left {
+            Some(ScalarRefImpl::List(left)) => {
+                if left.values_ref().len().to_i64().is_some() {
+                    Some(
+                        ListValue::new(
+                            left.values_ref()
+                                .into_iter()
+                                .enumerate()
+                                .filter(|(_, x)| x == &right)
+                                .map(|(idx, _)| {
+                                    Some(ScalarImpl::Int64((idx + 1).to_i64().unwrap()))
+                                })
+                                .collect(),
+                        )
+                        .into(),
+                    )
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ArrayPositionsExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let left_array = self.left.eval_checked(input).await?;
+        let right_array = self.right.eval_checked(input).await?;
+        let mut builder = self
+            .return_type
+            .create_array_builder(left_array.len() + right_array.len());
```

Comment:
https://github.com/risingwavelabs/risingwave/pull/9116#discussion_r1163891859

Actually, this is a signal that we should avoid copy-paste here, as fixes in one place should also be applied to other places. It seems the only core part is `evaluate` function, which is also why we have a function stored in some `FooExpression` and had to manually `impl Debug` to omit printing the function as part of the struct.

Going one step further, we may be able to leverage existing expression templates. Take a look at this:
https://github.com/risingwavelabs/risingwave/blob/84f4c689a0d9db80a5600fa520178789464513b9/src/expr/src/expr/expr_array_length.rs#L60-L63

We may be able to do something like:
```
 #[function("array_positions(list, *) -> list")] 
 fn array_positions<T: ScalarRef>(array: Option<ListRef<'_>>, element: Option<T>) -> Result<Option<ListValue>>
```

Feel free to come back with questions if the templates are too limited and cannot be used here.